ast/src/analyzed/mod.rs

```diff
@@ -87,6 +89,109 @@ impl<T> Analyzed<T> {
             })
             .sum()
     }
+
+    /// Removes the specified polynomials and updates the IDs of the other polynomials
+    /// so that they are contiguous again.
+    /// There must not be any reference to the removed polynomials left.
+    pub fn remove_polynomials(&mut self, to_remove: &BTreeSet<PolyID>) {
+        fn fold_replacement(
+            to_remove: &BTreeSet<PolyID>,
+            shift: u64,
+            mut replacements: BTreeMap<PolyID, PolyID>,
+            poly: PolyID,
+        ) -> (u64, BTreeMap<PolyID, PolyID>) {
+            if to_remove.contains(&poly) {
+                (shift + 1, replacements)
+            } else {
+                replacements.insert(
+                    poly,
+                    PolyID {
+                        id: poly.id - shift,
+                        ..poly
+                    },
+                );
+                (shift, replacements)
+            }
+        }
+
+        // TODO intermediate polys
+        let witness_replacements = self
+            .committed_polys_in_source_order()
+            .iter()
+            .fold(
+                (0, BTreeMap::new()),
+                |(shift, replacements), (poly, _def)| {
+                    fold_replacement(to_remove, shift, replacements, poly.into())
+                },
+            )
+            .1;
+        let fixed_replacements = self
+            .constant_polys_in_source_order()
+            .iter()
+            .fold(
+                (0, BTreeMap::new()),
+                |(shift, replacements), (poly, _def)| {
+                    fold_replacement(to_remove, shift, replacements, poly.into())
+                },
+            )
+            .1;
```

Comment:
Not sure it compiles, but to reduce repeated code, how do you like the following?
```rs
let [witness_replacements, fixed_replacements] = [
    self.committed_polys_in_source_order(),
    self.constant_polys_in_source_order()
].map(...);
```